File: navio_jupyter/navio_jupyter.py
import ipywidgets as widgets
from traitlets import Unicode, default, List, Integer, Dict

@widgets.register
class navio(widgets.DOMWidget):
    """An example widget."""
    _view_name = Unicode('NavioView').tag(sync=True)
    _model_name = Unicode('NavioModel').tag(sync=True)
    _view_module = Unicode('navio-jupyter').tag(sync=True)
    _model_module = Unicode('navio-jupyter').tag(sync=True)
    _view_module_version = Unicode('^0.0.1').tag(sync=True)
    _model_module_version = Unicode('^0.0.1').tag(sync=True)

    _view_height = Integer(600).tag(sync=True)
    _view_opts = Dict([]).tag(sync=True)
    _model_data = List([]).tag(sync=True)
    _model_selected = List([]).tag(sync=True)

    @default('layout')
    def _default_layout(self):
      return widgets.Layout(height=str(self._view_height)+'px', align_self='stretch')

    def set_data(self, js_data):
      self._model_data = js_data

    def set_opts(self, js_opts):
      self._view_opts = js_opts

    def set_height(self, js_height):
      self._view_height = js_height

    def get_selected(self):
      return self._model_selected

File: js/lib/navio-jupyter.js
var widgets = require("@jupyter-widgets/base");
var _ = require("lodash");
var navio = require("navio");


// Custom Model. Custom widgets models must at least provide default values
// for model attributes, including
//
//  - `_view_name`
//  - `_view_module`
//  - `_view_module_version`
//
//  - `_model_name`
//  - `_model_module`
//  - `_model_module_version`
//
//  when different from the base class.

// When serialiazing the entire widget state for embedding, only values that
// differ from the defaults will be specified.
var NavioModel = widgets.DOMWidgetModel.extend({
  defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
    _model_name : "NavioModel",
    _view_name : "NavioView",
    _model_module : "navio-jupyter",
    _view_module : "navio-jupyter",
    _model_module_version : "0.0.1",
    _view_module_version : "0.0.1",
    _model_data : [],
    _view_opts : {},
    _view_height : 600,
    _model_selected : [],
  })
});


// Custom View. Renders the widget model.
var NavioView = widgets.DOMWidgetView.extend({
  render: function() {
    var height = this.model.get("_view_height");
    this.nv = new navio(this.el, height);
    var data = this.model.get("_model_data");
    this.nv.data(data);
    // Initially all is selected
    this.model.set("_model_selected", data);
    this.nv.addAllAttribs();


    this.nv.updateCallback(this.updateSelected.bind(this));

    this.opts_changed();
    this.data_changed();
    this.model.on("change:data", this.data_changed, this);
    this.model.on("change:opts", this.opts_changed, this);
  },

  data_changed: function() {
    var data = this.model.get("_model_data");
    console.log("Navio changed data", data);
    this.nv.data(data);

    this.nv.update();
  },

  opts_changed: function() {
    var opts = this.model.get("_view_opts");
    console.log("Navio changed opts", opts);
    for(var o in opts) {
      if (opts.hasOwnProperty(o)) {
        this.nv[o] = opts[o];
      }
    }

    this.nv.update();
  },

  updateSelected: function(selected) {
    this.model.set("_model_selected", selected);
    this.model.save_changes();
  }
});


module.exports = {
  NavioModel : NavioModel,
  NavioView : NavioView
};
